### PR TITLE
implement built-in function `str_to_date`

### LINKED ISSUE #518: Implement built-in `STR_TO_DATE`
Function specification: https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_str-to-date

This built-in function offers more flexibility for parsing date values than the existing `DATETIME` function already present in this implementation.

### PR DESCRIPTION
**Pull request by cmoog**
Closes #518 

This PR implements the [STR_TO_DATE](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_str-to-date) MySQL function.

In places where the spec is ambiguous, I'm attempting to match the behavior of MySQL version 8 from my manual testing. I need to implement a few more parsers and add more test cases to cover the expected behavior.

```
go test  -cover github.com/dolthub/go-mysql-server/sql/parse/dateparse

ok  	github.com/dolthub/go-mysql-server/sql/parse/dateparse	0.163s	coverage: 89.3% of statements
```

cc @zachmu

### CODE REVIEW

File: sql/parse/dateparse/date.go [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+	// %T	Time, 24-hour (hh:mm:ss)
+	'T': parse24HourTimestamp,
+	'U': nil,
+	'u': nil,
+	'V': nil,
+	'v': nil,
+	'W': nil,
+	'w': nil,
+	'X': nil,
+	'x': nil,
```
Comment (cmoog) [inline sql/parse/dateparse/date.go:217]: TODO

**Comment by cmoog** [comment]
@zachmu it doesn't seem I can tag reviewers, but this is ready for a first pass while I finish up the final few format specifiers.

**Review by zachmu — COMMENTED** [review]
Overall this looks really good, thanks for the contribution!

The only things missing:

1) An engine test (e.g. enginetest/queries.go)
2) I think the time zone semantics are incorrect. Probably need all results in UTC for this to work reliably for everyone

**Comment by zachmu** [comment]
I missed the timezone setup in the test, I get it now :)

I'm going to go ahead and merge this in, please add an engine test as a followup. You can implement the missing week specifiers, or leave them as TODO for now. Those are pretty obscure for the purpose of date parsing.

**Pull request merged**